contracts/TaskTreasury/TaskTreasuryUpgradable.sol

```diff
@@ -0,0 +1,230 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.11;
+
+import {
+    EnumerableSet
+} from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import {
+    SafeERC20,
+    IERC20
+} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {
+    OwnableUpgradeable
+} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import {
+    ReentrancyGuardUpgradeable
+} from "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import {
+    Initializable
+} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+import {_transfer, ETH} from "../vendor/gelato/FGelato.sol";
+import {ITaskTreasury} from "../interfaces/ITaskTreasury.sol";
+
+contract TaskTreasuryUpgradable is
+    Initializable,
+    OwnableUpgradeable,
+    ReentrancyGuardUpgradeable
+{
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using SafeERC20 for IERC20;
+
+    address payable public immutable gelato;
+    ITaskTreasury public immutable oldTreasury;
+
+    mapping(address => mapping(address => uint256)) public userTokenBalance;
+    mapping(address => EnumerableSet.AddressSet) internal _tokenCredits;
+    EnumerableSet.AddressSet internal _whitelistedServices;
+
+    event FundsDeposited(
+        address indexed sender,
+        address indexed token,
+        uint256 indexed amount
+    );
+    event FundsWithdrawn(
+        address indexed receiver,
+        address indexed initiator,
+        address indexed token,
+        uint256 amount
+    );
+
+    event LogDeductFees(
+        address indexed user,
+        address indexed executor,
+        address indexed token,
+        uint256 fees,
+        address service
+    );
+
+    modifier onlyWhitelistedServices() {
+        require(
+            _whitelistedServices.contains(msg.sender),
+            "TaskTreasuryAccounting: onlyWhitelistedServices"
+        );
+        _;
+    }
+
+    constructor(address payable _gelato, address _oldTreasury) {
+        gelato = _gelato;
+        oldTreasury = ITaskTreasury(_oldTreasury);
+    }
+
+    receive() external payable {
```

Comment:
this function would let the implementation contract receive ether, the poxy already has a receive because you used `EIP173ProxyWithReceive` i'm not sure how to implement a custom receive function with hardhat-deploy. But usually the implementation contract shouldn't hold any funds.

Reply:
My thought was instead of hard coding in the proxy's `receive()` function, the proxy would delegate call the implementation's `receive()` to keep it backward compatible. If funds were sent to the implementation by accident, users could still withdraw them.

Reply:
nice i see what you did! Did you write a test case for that?

Reply:
added in c1056a1. Also realised sending ETH to implementation would revert in the first place